pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -97,6 +98,9 @@
     protected volatile Timeout batchReceiveTimeout = null;
     protected final Lock reentrantLock = new ReentrantLock();
 
+    // Only work when subscription type is Failover or Exclusive
+    protected final Lock incomingQueueLock;
```

Comment:
The `enqueueMessageAndCheckBatchReceive` method is always executed by the `internalPinnedExecutor` and the `redeliverUnacknowledgedMessages` method of `MultiTopicsConsumerImpl` is also executed by the `internalPinnedExecutor`. Is it possible to make sure the `redeliverUnacknowledgedMessages` method of the `ConsumerImpl` so that we don't need to introduce the lock here?

Reply:
BTW, enqueueMessageAndCheckBatchReceive is a hot path, maybe it will introduce performance degradation

Reply:
Make sense, I will try at it.

Reply:
@codelipenghui We must let the redeliver clear `incomingQueue` in the user thread, otherwise the redeliver will be meaningless, and you may still receive messages before the `incomingQueue` is cleared.

Reply:
> @codelipenghui We must let the redeliver clear incomingQueue in the user thread, otherwise the redeliver will be meaningless, and you may still receive messages before the incomingQueue is cleared.

Yes, So, Even if running in `internalPinned` thread, it still needs to be blocked. Please refer to the latest changes. @codelipenghui @congbobo184